service/controller/chart/resource/release/update.go

```diff
@@ -229,5 +245,96 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release does not have to be updated", desiredReleaseState.Name))
 	}
 
+	err = r.deleteRollbackAnnotation(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
 	return nil, nil
 }
+
+func (r *Resource) rollback(ctx context.Context, obj interface{}, currentStatus string) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	counts, ok := cr.GetAnnotations()[annotation.RollbackCount]
+	var rollbackCounts int
+	if !ok {
+		rollbackCounts = 0
+	} else {
+		rollbackCounts, err = strconv.Atoi(counts)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if rollbackCounts > r.maxRollback {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and cannot be updated despite of %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
+		return nil
+	}
+
+	if currentStatus == helmclient.StatusPendingInstall {
+		err = r.helmClient.DeleteRelease(ctx, key.Namespace(cr), key.ReleaseName(cr))
```

Comment:
We should have logging before and after this. As deleting the release changes the state.